src/FastConv/FastConv.cpp

```diff
@@ -0,0 +1,130 @@
+
+#include "FastConv.h"
+
+CFastConv::CFastConv( void )
+{
+}
+
+CFastConv::~CFastConv( void )
+{
+    reset();
+}
+
+Error_t CFastConv::init(float *pfImpulseResponse, int iLengthOfIr, int iBlockLength)
+{
+
+    m_ImpulseResponse = pfImpulseResponse;
+    m_blockLength = iBlockLength;
+    m_IRlen = iLengthOfIr;
+    m_pcRingBuff = new CRingBuffer<float>(iLengthOfIr);
+    CFft::createInstance(fft);
+    fft->initInstance(iBlockLength);
+
+    return Error_t::kNoError;
+}
+
+Error_t CFastConv::reset()
+{
+    m_ImpulseResponse = 0;
+//    delete [] m_pcRingBuff;
+    m_ImpulseResponse = 0;
+    m_pcRingBuff = 0;
+
+    return Error_t::kNoError;
+}
+
+Error_t CFastConv::process (float* pfOutputBuffer, const float *pfInputBuffer, int iLengthOfBuffers )
+{
+
+    std::memset(pfOutputBuffer, 0, iLengthOfBuffers);
+    for (int i = 0; i < iLengthOfBuffers; i ++)
+    {
+        m_pcRingBuff->setReadIdx(m_pcRingBuff->getWriteIdx() + 1);
+        m_pcRingBuff->putPostInc(pfInputBuffer[i]);
+
+        for (int j = m_IRlen - 1; j >= 0; j--)
+        {
+            pfOutputBuffer[i] += m_ImpulseResponse[j] * m_pcRingBuff->getPostInc();
+        }
+
+    }
+    return Error_t::kNoError;
+}
+
+Error_t CFastConv::flushBuffer(float* pfOutputBuffer)
+{
+
+    return Error_t::kNoError;
+}
+
+void CFastConv::processFrequencyDomain(float *pfOutputBuffer, const float* pfInputBuffer, int iLengthofBuffers)
+{
+    long seqLen = sizeof(*pfInputBuffer)/sizeof (pfInputBuffer[0]) + m_blockLength - 1;
```

Comment:
All of this memory allocation can be made in your init, just use the eCompMode as an if and allocate the required memory for either freqdomain process or time domain. Some people have created a base class and child classes for each of these just like our combfilter assignment, that’s also an option to look into